data/basics/teal.go

```diff
@@ -0,0 +1,298 @@
+func (vd *ValueDelta) ToTealValue() (value TealValue, ok bool) {
+	switch vd.Action {
+	case SetBytesAction:
+		value.Type = TealBytesType
+		value.Bytes = vd.Bytes
+		ok = true
+	case SetUintAction:
+		value.Type = TealUintType
+		value.Uint = vd.Uint
+		ok = true
+	case DeleteAction:
+		value.Type = TealUintType
+		ok = false
+	default:
+		value.Type = TealUintType
+		ok = false
+	}
+	return value, ok
+}
+
+// StateDelta is a map from key/value store keys to ValueDeltas, indicating
+// what should happen for that key
+//msgp:allocbound StateDelta config.MaxStateDeltaKeys
+type StateDelta map[string]ValueDelta
+
+// Equal checks whether two StateDeltas are equal. We don't check for nilness
+// equality because an empty map will encode/decode as nil. So if our generated
+// map is empty but not nil, we want to equal a decoded nil off the wire.
+func (sd StateDelta) Equal(o StateDelta) bool {
+	// Lengths should be the same
+	if len(sd) != len(o) {
+		return false
+	}
+	// All keys and deltas should be the same
+	for k, v := range sd {
+		if o[k] != v {
+			return false
+		}
+	}
+	return true
+}
+
+// Valid checks whether the keys and values in a StateDelta conform to the
+// consensus parameters' maximum lengths
+func (sd StateDelta) Valid(proto *config.ConsensusParams) error {
+	for key, delta := range sd {
+		if len(key) > proto.MaxAppKeyLen || proto.MaxAppKeyLen == 0 {
+			return fmt.Errorf("key too long: length was %d, maximum is %d", len(key), proto.MaxAppKeyLen)
+		}
+		switch delta.Action {
+		case SetBytesAction:
+			if len(delta.Bytes) > proto.MaxAppBytesValueLen {
+				return fmt.Errorf("cannot set value for key 0x%x, too long: length was %d, maximum is %d", key, len(delta.Bytes), proto.MaxAppBytesValueLen)
+			}
+		case SetUintAction:
+		case DeleteAction:
+		default:
+			return fmt.Errorf("unknown delta action")
+		}
+	}
+	return nil
+}
+
+// EvalDelta stores StateDeltas for an application's global key/value store, as
+// well as StateDeltas for some number of accounts holding local state for that
+// application
+type EvalDelta struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	GlobalDelta StateDelta `codec:"gd"`
+
+	// When decoding EvalDeltas, the integer key represents an offset into
+	// [txn.Sender, txn.Accounts[0], txn.Accounts[1], ...]
+	LocalDeltas map[uint64]StateDelta `codec:"ld,allocbound=config.MaxEvalDeltaAccounts"`
+}
+
+// Equal compares two EvalDeltas and returns whether or not they are
+// equivalent. It does not care about nilness equality of LocalDeltas,
+// because the msgpack codec will encode/decode an empty map as nil, and we want
+// an empty generated EvalDelta to equal an empty one we decode off the wire.
+func (ed EvalDelta) Equal(o EvalDelta) bool {
+	// GlobalDeltas must be equal
+	if !ed.GlobalDelta.Equal(o.GlobalDelta) {
+		return false
+	}
+
+	// LocalDeltas length should be the same
+	if len(ed.LocalDeltas) != len(o.LocalDeltas) {
```

Comment:
can you move this check first ? it's much faster than map items comparison.

Reply:
done